syntax = "proto3";

import "google/protobuf/any.proto";
import "google/protobuf/timestamp.proto";

message CloudEvent {

  string id = 1;
  string source = 2;
  string spec_version = 3;
  string type = 4;

  map<string, CloudEventAttributeValue> attributes = 5;

  oneof  data {
    bytes binary_data = 6;
    string text_data = 7;
    google.protobuf.Any proto_data = 8;
  }

  message CloudEventAttributeValue {
    oneof attr {
      bool ce_boolean = 1;
      int32 ce_integer = 2;
      string ce_string = 3;
      bytes ce_bytes = 4;
      string ce_uri = 5;
      string ce_uri_ref = 6;
      google.protobuf.Timestamp ce_timestamp = 7;
    }
  }
}

message CloudEventBatch {
  repeated CloudEvent events = 1;
}